src/mmds/src/token.rs

```diff
@@ -0,0 +1,563 @@
+    }
+
+    /// Create a new AES-GCM cipher entity.
+    fn create_cipher(entropy_pool: &mut File) -> Result<Aes256Gcm, Error> {
+        // Randomly generate a 256-bit key to be used for encryption/decryption purposes.
+        let mut key = [0u8; KEY_LEN];
+        entropy_pool
+            .read_exact(&mut key)
+            .map_err(Error::EntropyPool)?;
+
+        // Create cipher entity to handle encryption/decryption.
+        Ok(Aes256Gcm::new(Key::from_slice(&key)))
+    }
+
+    /// Make sure to reinitialize the cipher under a new key before reaching
+    /// a count of 2^32 encrypted tokens under the same cipher entity.
+    fn check_encryption_count(&mut self) -> Result<(), Error> {
+        // Make sure no more than 2^32 - 1 tokens are encrypted under
+        // the same encryption key.
+        if self.num_encrypted_tokens == u32::MAX {
+            // Reinitialize the cipher entity under a new key when limit is exceeded.
+            // As a result, all valid tokens created under the previous key are invalidated.
+            // By design, we don't retain the cipher used to encrypt previous tokens,
+            // because reaching the limit is very unlikely and should not happen under
+            // healthy interactions with MMDS. However, if it happens, we expect the
+            // customer code to have a retry mechanism in place and regenerate the
+            // session token if the one provided is invalid.
+            self.cipher = TokenAuthority::create_cipher(&mut self.entropy_pool)?;
+            // Reset encrypted tokens count.
+            self.num_encrypted_tokens = 0;
+            warn!(
+                "MMDS's token authority entity has been reseeded. \
+                All previously created tokens are now invalid."
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Validate the token time to live against bounds.
+    fn check_ttl(ttl_seconds: u32) -> bool {
+        MIN_TOKEN_TTL_SECONDS <= ttl_seconds && ttl_seconds <= MAX_TOKEN_TTL_SECONDS
+    }
+
+    /// Compute expiry time in seconds by adding the time to live provided
+    /// to the current time measured in seconds.
+    fn compute_expiry(ttl_seconds: u32) -> u64 {
+        // Get time elapsed since UNIX_EPOCH.
+        let now_as_duration = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("System time is before UNIX_EPOCH.");
+        // Convert elapsed time into seconds.
+        let ttl_duration = Duration::from_secs(u64::from(ttl_seconds));
+
+        // Compute expiry by adding ttl value to current time in seconds.
+        // This addition is safe because ttl is verified beforehand and
+        // can never be more than 6h.
+        let expiry_duration = now_as_duration.add(ttl_duration);
+
+        expiry_duration.as_secs()
+    }
+}
+
+/// Structure for token information.
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+struct Token {
+    version: u16,
```

Comment:
Why do we need the token version though?
Do we envision adding new fields to this token struct and supporting older versions of the struct in newer Firecracker versions?

Reply:
As we've chatted offline, this would only be useful only for loading snapshots older guests in a newer FC versions (which have a new version of the MMDS token). But this wouldn't work since the `TokenAuthority` will be reinitialised and all tokens invalidated.
The only other possible use case I can think of would be for live migration (but we should tackle this when we have some clear plans for it)

Reply:
I removed the `version` field inside the Token struct, as we discussed.